File: functions/Petshop.Shared/Messages/ServiceBusMessages.cs
namespace Petshop.Shared.Messages;

/// <summary>
/// Mensagem enviada quando um pedido é confirmado.
/// O Catalog Service consome para deduzir estoque.
/// </summary>
public class OrderConfirmedMessage
{
    public long PedidoId { get; set; }
    public long ClienteId { get; set; }
    public DateTime DataPedido { get; set; }
    public List<OrderItemMessage> Itens { get; set; } = new();
}

public class OrderItemMessage
{
    public long ProdutoId { get; set; }
    public int Quantidade { get; set; }
}

/// <summary>
/// Mensagem enviada quando um pedido é cancelado.
/// O Catalog Service consome para restaurar estoque.
/// </summary>
public class OrderCancelledMessage
{
    public long PedidoId { get; set; }
    public long ClienteId { get; set; }
    public List<OrderItemMessage> Itens { get; set; } = new();
}

/// <summary>
/// Mensagem enviada quando o status de um pedido muda.
/// Pode ser consumida por um serviço de notificação.
/// </summary>
public class OrderStatusChangedMessage
{
    public long PedidoId { get; set; }
    public long ClienteId { get; set; }
    public string ClienteEmail { get; set; } = string.Empty;
    public string StatusAnterior { get; set; } = string.Empty;
    public string StatusNovo { get; set; } = string.Empty;
    public DateTime DataAlteracao { get; set; }
}

/// <summary>
/// Mensagem enviada quando um agendamento é criado.
/// Pode ser consumida por um serviço de notificação.
/// </summary>
public class AppointmentCreatedMessage
{
    public long AgendamentoId { get; set; }
    public long ClienteId { get; set; }
    public string ClienteEmail { get; set; } = string.Empty;
    public string ClienteNome { get; set; } = string.Empty;
    public string PetNome { get; set; } = string.Empty;
    public DateTime DataAgendamento { get; set; }
    public TimeSpan Horario { get; set; }
    public List<string> Servicos { get; set; } = new();
}

/// <summary>
/// Mensagem para lembretes de agendamento.
/// Enviada pelo Timer Trigger 24h antes.
/// </summary>
public class AppointmentReminderMessage
{
    public long AgendamentoId { get; set; }
    public string ClienteEmail { get; set; } = string.Empty;
    public string ClienteNome { get; set; } = string.Empty;
    public string PetNome { get; set; } = string.Empty;
    public DateTime DataAgendamento { get; set; }
    public TimeSpan Horario { get; set; }
    public string MetodoAtendimento { get; set; } = string.Empty;
}

/// <summary>
/// Mensagem para alertas de estoque baixo.
/// Enviada pelo Timer Trigger do Catalog Service.
/// </summary>
public class LowStockAlertMessage
{
    public long ProdutoId { get; set; }
    public string ProdutoNome { get; set; } = string.Empty;
    public int QuantidadeAtual { get; set; }
    public int LimiteMinimo { get; set; }
    public DateTime DataAlerta { get; set; }
}

/// <summary>
/// Mensagem para dedução de estoque.
/// Consumida pelo Catalog Service.
/// </summary>
public class StockDeductionMessage
{
    public long PedidoId { get; set; }
    public List<StockItemMessage> Items { get; set; } = new();
}

/// <summary>
/// Mensagem para restauração de estoque.
/// Consumida pelo Catalog Service quando pedido é cancelado.
/// </summary>
public class StockRestoreMessage
{
    public long PedidoId { get; set; }
    public List<StockItemMessage> Items { get; set; } = new();
}

public class StockItemMessage
{
    public long ProdutoId { get; set; }
    public int Quantidade { get; set; }
}

/// <summary>
/// Mensagem para confirmação de agendamento pelo cliente.
/// </summary>
public class SchedulingConfirmationMessage
{
    public long AgendamentoId { get; set; }
    public long ClienteId { get; set; }
    public bool Confirmado { get; set; }
    public string? MotivoRecusa { get; set; }
}

File: functions-java/petshop-functions-shared/src/main/java/com/petshop/functions/shared/messages/ServiceBusMessages.java
package com.petshop.functions.shared.messages;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

/**
 * Classes de mensagens para Azure Service Bus
 */
public final class ServiceBusMessages {

    private ServiceBusMessages() {}

    /**
     * Mensagem para dedução de estoque
     */
    public record StockDeductionMessage(
            Long pedidoId,
            List<StockItem> items,
            LocalDateTime timestamp
    ) implements Serializable {
        public record StockItem(Long produtoId, int quantidade) implements Serializable {}
    }

    /**
     * Mensagem para restauração de estoque (cancelamento)
     */
    public record StockRestoreMessage(
            Long pedidoId,
            List<StockDeductionMessage.StockItem> items,
            String reason,
            LocalDateTime timestamp
    ) implements Serializable {}

    /**
     * Mensagem de confirmação de pedido
     */
    public record OrderConfirmedMessage(
            Long pedidoId,
            Long clienteId,
            BigDecimal valorTotal,
            String status,
            LocalDateTime timestamp
    ) implements Serializable {}

    /**
     * Mensagem de atualização de status de pedido
     */
    public record OrderStatusChangedMessage(
            Long pedidoId,
            String statusAnterior,
            String statusNovo,
            LocalDateTime timestamp
    ) implements Serializable {}

    /**
     * Mensagem de agendamento criado
     */
    public record SchedulingCreatedMessage(
            Long agendamentoId,
            Long clienteId,
            Long petId,
            Long servicoId,
            LocalDateTime dataHora,
            LocalDateTime timestamp
    ) implements Serializable {}

    /**
     * Mensagem de agendamento cancelado
     */
    public record SchedulingCancelledMessage(
            Long agendamentoId,
            String motivo,
            LocalDateTime timestamp
    ) implements Serializable {}

    /**
     * Mensagem de cliente registrado
     */
    public record CustomerRegisteredMessage(
            Long clienteId,
            Long usuarioId,
            String nome,
            String email,
            LocalDateTime timestamp
    ) implements Serializable {}

    /**
     * Mensagem de notificação genérica
     */
    public record NotificationMessage(
            String tipo,
            String destinatario,
            String assunto,
            String conteudo,
            LocalDateTime timestamp
    ) implements Serializable {}
}
